File: Asset Browser UI Component/Code/PSD Extraction/PSD Extraction/ViewController.m
//
//  ViewController.m
//  PSD Extraction
//
//  Copyright (c) 2015 Adobe Systems Incorporated. All rights reserved.
//
//  Permission is hereby granted, free of charge, to any person obtaining a
//  copy of this software and associated documentation files (the "Software"),
//  to deal in the Software without restriction, including without limitation
//  the rights to use, copy, modify, merge, publish, distribute, sublicense,
//  and/or sell copies of the Software, and to permit persons to whom the
//  Software is furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
//  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
//  DEALINGS IN THE SOFTWARE.
//

#import <AdobeCreativeSDKCore/AdobeCreativeSDKCore.h>
#import <AdobeCreativeSDKAssetModel/AdobeCreativeSDKAssetModel.h>
#import <AdobeCreativeSDKAssetUX/AdobeCreativeSDKAssetUX.h>

#import "ViewController.h"
#import "LayerTableViewCell.h"

//warning Change the Client ID and Secret to match the values provided by the app registration website (creativesdk.adobe.com)

static NSString * const kAppClientId = @"changeme";
static NSString * const kAppClientSecret = @"changeme";

@interface ViewController () <UITableViewDataSource, UITableViewDelegate>

@property (weak, nonatomic) IBOutlet UIButton *loginButton;
@property (weak, nonatomic) IBOutlet UILabel *psdFileNameLabel;
@property (weak, nonatomic) IBOutlet UITableView *tableView;

@property (strong, nonatomic) AdobeAssetPSDFile *psdFile;
@property (strong, nonatomic) NSArray *selectedLayers;

@end

@implementation ViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    [[AdobeUXAuthManager sharedManager] setAuthenticationParametersWithClientID:kAppClientId
                                                                   clientSecret:kAppClientSecret
                                                                   enableSignUp:NO];
    
    // layoutMargins doesn't exist on iOS 7 so check before calling it.
    if ([self.tableView respondsToSelector:@selector(setLayoutMargins:)])
    {
        self.tableView.layoutMargins = UIEdgeInsetsZero;
    }
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    
    if ([AdobeUXAuthManager sharedManager].isAuthenticated)
    {
        [self.loginButton setTitle:NSLocalizedString(@"Log Out", @"Logout button")
                          forState:UIControlStateNormal];
    }
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark UI Actions

- (IBAction)loginButtonTouchUpInside
{
    if ([AdobeUXAuthManager sharedManager].isAuthenticated)
    {
        [[AdobeUXAuthManager sharedManager] logout:^{
            
            [self.loginButton setTitle:NSLocalizedString(@"Log In", @"Login button") forState:UIControlStateNormal];
            
            self.psdFileNameLabel.text = nil;
            self.selectedLayers = nil;
            [self.tableView reloadData];
            self.tableView.hidden = YES;
            
        } onError:^(NSError *error) {
            
            NSLog(@"Error in logout: %@", error);
        }];
    }
    else
    {
        [[AdobeUXAuthManager sharedManager] login:self onSuccess:^(AdobeAuthUserProfile *profile) {
            
            [self.loginButton setTitle:NSLocalizedString(@"Log Out", "Logout button")
                              forState:UIControlStateNormal];
            
        } onError:^(NSError *error) {
            
            NSLog(@"Error in login: %@", error);
        }];
    }
}

- (IBAction)showAssetBrowserTouchUpInside
{
    // Exclude all other data sources. Only allow the "Files" datasource
    AdobeAssetDataSourceFilter *dataSourceFilter = [[AdobeAssetDataSourceFilter alloc] initWithDataSources:@[AdobeAssetDataSourceFiles]
                                                                                                filterType:AdobeAssetDataSourceFilterInclusive];
    
    // Exclude all other file types, other than PSD files.
    AdobeAssetMIMETypeFilter *mimeTypeFilter = [[AdobeAssetMIMETypeFilter alloc] initWithMIMETypes:@[kAdobeMimeTypePhotoshop]
                                                                                        filterType:AdobeAssetMIMETypeFilterTypeInclusion];
    
    // Create a new configuration object that can be used to customize the Asset Browser's behavior
    // or appearance.
    AdobeUXAssetBrowserConfiguration *configuration = [AdobeUXAssetBrowserConfiguration new];
    
    // Set all the options
    configuration.dataSourceFilter = dataSourceFilter;
    configuration.mimeTypeFilter = mimeTypeFilter;
    configuration.options = EnablePSDLayerExtraction | EnableMultiplePSDLayerSelection;
    
    // Call the Asset Browser and pass the configuration options
    [[AdobeUXAssetBrowser sharedBrowser] popupFileBrowserWithParent:self configuration:configuration onSuccess:^(NSArray *itemSelections) {
        
        // Grab the last item that was selected.
        AdobeSelectionAsset *itemSelection = itemSelections.lastObject;
        
        // Make sure it's a PSD file.
        if (IsAdobeSelectionAssetPSDFile(itemSelection))
        {
            // We know the selected item is a PSD file so we can safely cast it to the specific type.
            AdobeSelectionAssetPSDFile *psdSelection = (AdobeSelectionAssetPSDFile *)itemSelection;
            
            // Grab the actual Asset file instance.
            AdobeAssetPSDFile *psdFile = (AdobeAssetPSDFile *)psdSelection.selectedItem;
            
            self.psdFileNameLabel.text = [NSString stringWithFormat:@"PSD File: %@", psdFile.name];
            self.psdFile = psdFile;
            
            // Also grab all the selected layers
            AdobePSDLayerSelectionArray *layerSelections = psdSelection.layerSelections;
            
            NSMutableArray *selectedLayers = [NSMutableArray arrayWithCapacity:psdSelection.layerSelections.count];
            
            for (AdobeSelectionPSDLayer *psdLayerSelection in layerSelections)
            {
                [selectedLayers addObject:psdLayerSelection.layer];
            }
            
            self.selectedLayers = selectedLayers;
            
            [self.tableView reloadData];
            self.tableView.hidden = NO;
        }
        else
        {
            
        }
        
    } onError:^(NSError *error) {
        
        NSLog(@"An error occurred: %@", error);
    }];
}

#pragma mark UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.selectedLayers.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    LayerTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
    
    if (cell == nil)
    {
        cell = [LayerTableViewCell new];
    }
    
    AdobePSDLayerNode *layer = self.selectedLayers[indexPath.row];
    
    cell.layerId = layer.layerId;
    
    NSString *type = nil;
    
    switch (layer.type)
    {
        case AdobePSDLayerNodeTypeRGBPixels:
        {
            type = @"RGP";
            break;
        }
        case AdobePSDLayerNodeTypeSolidColor:
        {
            type = @"Solid color";
            break;
        }
        case AdobePSDLayerNodeTypeGroup:
        {
            type = @"Layer Group";
            break;
        }
        case AdobePSDLayerNodeTypeAdjustment:
        {
            type = @"Adjustment";
            break;
        }
        case AdobePSDLayerNodeTypeUnknown:
        {
            type = @"Unknown";
            break;
        }
    }
    
    // Create a temporary location for the result string.
    NSString *layerInformation = [NSString stringWithFormat:@"Selected Layer: %li\nLayer name: %@\nType: %@\nLayer ID: %@\nLayer Index: %li\nVisible: %@",
                                  (long)indexPath.row, layer.name, type, layer.layerId, (long)layer.layerIndex, layer.visible ? @"Yes" : @"No"];
    
    cell.layerInformation = layerInformation;
    
    [self.psdFile getRenditionForLayer:layer.layerId
                         withLayerComp:nil
                              withType:AdobeAssetFileRenditionTypePNG
                              withSize:CGSizeMake(120, 0)
                          withPriority:NSOperationQueuePriorityHigh
                            onProgress:NULL
                          onCompletion:^(NSData *imageData, BOOL fromCache)
    {
        if ([cell.layerId isEqualToNumber:layer.layerId])
        {
            UIImage *image = [UIImage imageWithData:imageData];
            
            NSLog(@"%@", NSStringFromCGSize(image.size));
            
            cell.thumbnailImage = image;
        }
    }
                        onCancellation:^
    {
        NSLog(@"Layer rendition cancelled for layer: %@", layer);
    }
                               onError:^(NSError *error)
    {
        NSLog(@"An error occured when fetching the rendition for layer: %@", layer);
    }];
    
    return cell;
}

@end

File: Asset Browser UI Component/Code/PSD Extraction/Swift/PSD Extraction/ViewController.swift
/*
 * Copyright (c) 2016 Adobe Systems Incorporated. All rights reserved.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 */

import UIKit

class ViewController: UIViewController
{
    // TODO: Please update the client ID and secret values to match the ones provided by 
    // creativesdk.com
    private let kCreativeSDKClientId = "Change me"
    private let kCreativeSDKClientSecret = "Change me"
    private let kCreativeSDKRedirectURLString = "Change me"
    
    @IBOutlet private weak var psdFileNameLabel: UILabel!
    @IBOutlet private weak var tableView: UITableView!
    
    var psdFile: AdobeAssetPSDFile?
    var selectedLayers: Array<AdobePSDLayerNode>!
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        // Set the client ID and secret values so the CSDK can identify the calling app. The three
        // specified scopes are required at a minimum.
        AdobeUXAuthManager.sharedManager().setAuthenticationParametersWithClientID(kCreativeSDKClientId,
                                                                                   clientSecret: kCreativeSDKClientSecret,
                                                                                   additionalScopeList: [
                                                                                    AdobeAuthManagerUserProfileScope,
                                                                                    AdobeAuthManagerEmailScope,
                                                                                    AdobeAuthManagerAddressScope])
        
        // Also set the redirect URL, which is required by the CSDK authentication mechanism.
        AdobeUXAuthManager.sharedManager().redirectURL = NSURL(string: kCreativeSDKRedirectURLString)
        
        // Reset the table view margins
        tableView.layoutMargins = UIEdgeInsetsZero
    }
    
    @IBAction
    func pickPSDFileButtonTouchUpInside()
    {
        // Only show the Files datasource.
        let dataSourceFilter = AdobeAssetDataSourceFilter(dataSources: [AdobeAssetDataSourceFiles], filterType: .Inclusive)
        
        // Only allow PSD file to be selected.
        let mimeTypeFilter = AdobeAssetMIMETypeFilter(MIMETypes: [kAdobeMimeTypePhotoshop], filterType: .Inclusion)
        
        // Create the configuration object and specify that PSD layer extraction (and multiple 
        // selection of those layers) should be enabled.
        let configuration = AdobeUXAssetBrowserConfiguration()
        configuration.dataSourceFilter = dataSourceFilter
        configuration.mimeTypeFilter = mimeTypeFilter
        configuration.options = AdobeUXAssetBrowserOption.EnablePSDLayerExtraction.rawValue | AdobeUXAssetBrowserOption.EnableMultiplePSDLayerSelection.rawValue
        
        // Create an instance of the Asset Browser view controller
        let assetBrowserViewController = AdobeUXAssetBrowserViewController(configuration: configuration, delegate: self)
        
        // Present the Asset Browser view controller
        self.presentViewController(assetBrowserViewController, animated: true, completion: nil)
    }
}

// MARK: - UITableViewDataSource
extension ViewController: UITableViewDataSource
{
    func numberOfSectionsInTableView(tableView: UITableView) -> Int
    {
        return 1
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int
    {
        return selectedLayers?.count ?? 0
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell
    {
        let cell = tableView.dequeueReusableCellWithIdentifier("Cell", forIndexPath: indexPath) as! LayerTableViewCell
        
        let layer = selectedLayers[indexPath.row]
        var type = ""
        
        switch layer.type
        {
            case AdobePSDLayerNodeType.RGBPixels:
                type = "RGB"
            
            case AdobePSDLayerNodeType.SolidColor:
                type = "Solid Color"
                
            case AdobePSDLayerNodeType.Group:
                type = "Layer Group"
                
            case AdobePSDLayerNodeType.Adjustment:
                type = "Adjustment"
            
            default:
                type = "Unknown"
        }
        
        cell.layerId = layer.layerId
        cell.layerInformation = "Selected Layer: \(indexPath.row)\n" +
            "Layer name: \(layer.name)\n" +
            "Type: \(type)\n" +
            "Layer ID: \(layer.layerId)\n" +
            "Layer Index: \(layer.layerIndex)\n" +
            "Visible: \(layer.visible)"
        
        psdFile?.downloadRenditionForLayerID(layer.layerId,
            layerCompID: nil,
            renditionType: .PNG,
            dimensions: CGSize(width: 120, height: 0),
            requestPriority: .High,
            progressBlock: nil,
            successBlock:
            {
                (imageData: NSData!, fromCache: Bool) -> Void in
                
                if cell.layerId == layer.layerId
                {
                    let image = UIImage(data: imageData)
                    
                    cell.thumbnailImage = image
                }
            },
            cancellationBlock:
            {
                print("Layer rendition cancelled for layer: \(layer)")
            },
            errorBlock:
            {
                (error: NSError!) -> Void in
                
                print("An error occurred when fetching the rendition for layer: \(layer)")
            }
         )
        
        return cell
    }
}

// MARK: - AdobeUXAssetBrowserViewControllerDelegate
extension ViewController : AdobeUXAssetBrowserViewControllerDelegate
{
    func assetBrowserDidSelectAssets(itemSelections: [AdobeSelectionAsset])
    {
        // Dismiss the Asset Browser view controller.
        self.dismissViewControllerAnimated(true, completion: nil)
        
        // Make sure something was selected.
        let itemSelection = itemSelections.first
        
        // Make sure the selected asset is a PSD selection.
        guard let psdSelection = itemSelection as? AdobeSelectionAssetPSDFile else
        {
            return
        }
        
        // Now grab the actual PSD item that was selected.
        guard let psdFile = psdSelection.selectedItem as? AdobeAssetPSDFile else
        {
            return
        }
        
        self.psdFileNameLabel.text = "PSD File: \(psdFile.name)"
        self.psdFile = psdFile
        
        // Grab all the selected layers
        let layerSelections = psdSelection.layerSelections
        var selectedLayers: Array<AdobePSDLayerNode> = []
        
        for psdLayerSelection in layerSelections
        {
            selectedLayers.append((psdLayerSelection as! AdobeSelectionPSDLayer).layer)
        }
        
        self.selectedLayers = selectedLayers
        
        // Reload the table view data
        self.tableView.reloadData()
        self.tableView.hidden = false
    }
    
    func assetBrowserDidEncounterError(error: NSError)
    {
        // Dismiss the Asset Browser view controller.
        self.dismissViewControllerAnimated(true, completion: nil)
        
        print("An error occurred: \(error)")
    }
    
    func assetBrowserDidClose()
    {
        print("The user closed the Asset Browser without choosing any assets.")
    }
}
